std/prelude.glu

```diff
@@ -18,54 +15,180 @@ let unwrap_err res : Result e a -> e =
     | Ok _ -> error "Result was an Ok"
     | Err x -> x
 
-/// `Monoid m` represents an associative operation on `m` an which has an identity.
+/// Boolean 'not'
+let not x : Bool -> Bool = if x then False else True
+
+/// Boolean 'exclusive or'
+let xor x y : Bool -> Bool -> Bool = if x then not y else y
+
+/// `Semigroup a` represents an associative operation on `a`.
 /// This means the following laws must hold:
 ///
-/// * `append x empty == x`
-/// * `append empty x == x`
-/// * `append x (append y z) == append (append x y) z`
-type Monoid m = {
-    /// # Note
-    ///
-    /// * Known as `mappend` in Haskell
-    append : m -> m -> m,
+/// * `forall x . append x (append y z) == append (append x y) z`
+type Semigroup a = {
```

Comment:
Something which has been bugging me is where these types and implementations should live? The prelude has become a bit of a kitchen sink of implementions whereas it should probably just reexport most if its contents from elsewhere.

Perhaps `SemiGroup`, `Monad` etc should each live in their own file? Or if that is overkill then maybe. But then where should the implementations live (`Bool`, `Option` etc)? Should they be in `std/semigroup.glu` or in `std/bool.glu`. If they live in `std/bool.glu` (ie they live the the type _implementing_) then perhaps `std/semigroup.glu` is to small of a file to warrants its own existence and we should instead have `std/functional.glu` as a module for all these types.

Should probably open an issue to work these things out since I don't think it needs to be fixed in this PR!